thirdeye/thirdeye-frontend/app/pods/rootcause/route.js

```diff
@@ -1,21 +1,105 @@
 
-export default Ember.Route.extend({
-  setupController: function (controller, model) {
+const queryParamsConfig = {
+  refreshModel: false,
+  replace: false
+};
+
+/**
+ * Helper function that checks if a query param 
+ * key/value pair is valid
+ * @param {*} key   - query param key
+ * @param {*} value - query param value
+ * @return {Boolean}
+ */
+const isValid = (key, value) => { 
+  switch(key) {
+    case 'granularity':
+      return ['MINUTES', 'HOURS', 'DAYS'].includes(value);
+    case 'filters':
+      return value && value.length;
+    default:
+      return moment(+value).isValid();
+  }
+};
+
+
+export default Ember.Route.extend(AuthenticatedRouteMixin, {
+  queryParams: {
+    granularity: queryParamsConfig,
+    filters: queryParamsConfig,
+    compareMode: queryParamsConfig,
+    analysisRangeStart: queryParamsConfig,
+    analysisRangeEnd: queryParamsConfig,
+    anomalyRangeStart: queryParamsConfig,
+    anomalyRangeEnd: queryParamsConfig
+  },
+
+  model(params) {
+    const { rootcauseId: id } = params;
+
+    // TODO handle error better
+    if (!id) { return; }
+
+    return RSVP.hash({
+      granularityOptions: fetch(`/data/agg/granularity/metric/${id}`).then(res => res.json()),
+      filterOptions: fetch(`/data/autocomplete/filters/metric/${id}`).then(res => res.json()),
+      maxTime: fetch(`/data/maxDataTime/metricId/${id}`).then(res => res.json()),
+      compareModeOptions: ['WoW', 'Wo2W', 'Wo3W', 'Wo4W'],
+      id
+    });
+  },
+
+  afterModel(model, transition) {
+
+    const defaultParams = {
+      filters: JSON.stringify({}),
+      granularity: model.granularityOptions[0],
+      anomalyRangeStart: moment().subtract(1, 'day').valueOf(),
+      anomalyRangeEnd: model.maxTime,
+      analysisRangeStart: moment().subtract(1, 'week').valueOf(), 
+      analysisRangeEnd: model.maxTime,
+      compareMode: 'WoW'
+    };
+    let { queryParams } = transition;
+    // TODO: write utils functions that checks key values in queryParams
+    //       so only valid strings are authorized
+    const validParams = Object.keys(queryParams)
+      .filter((param) => {
+        const value = queryParams[param];
+        return value && isValid(param, value);
+      })
+      .reduce((hash, key) => {
+        hash[key] = queryParams[key];
+        return hash;
+      }, {});
+    Object.assign(
+      model, 
+      { queryParams: { ...defaultParams, ...validParams }}
+    );
+  },
+
+  setupController(controller, model) {
     this._super(...arguments);
 
-    // TODO get initial attributes from query params
     controller.setProperties({
       selectedUrns: new Set(['thirdeye:metric:194592', 'thirdeye:event:holiday:2712391']),
       invisibleUrns: new Set(),
       filteredUrns: new Set(),
       hoverUrns: new Set(),
-      context: { urns, anomalyRange, baselineRange, analysisRange }
+      context: { urns, anomalyRange, baselineRange, analysisRange },
+      ...model.queryParams
```

Comment:
The controller should only handle the context object. All the complexity of converting between query-params and context, and display options and context should be hidden in the route and the components - i.e. "context" is the interface.

Reply:
In order to use queryparams, those needs to be defined in both the route and controller, because the queryparams are only stored in a one level object, we would need some sort of conversion. Since the `queryparams` only update the url but not the model, the model hooks and setupController are not called. Therefore the only place to have this conversion logic is in the setupController on init and the `onChange` handler passed into the component. 
Let me know if you want to handle this another way